SharpPcap/LibPcap/LibPcapSafeNativeMethods.Resolver.cs

```diff
@@ -0,0 +1,114 @@
+ * Copyright 2008-2009 Phillip Lemon <lucidcomms@gmail.com>
+ * Copyright 2008-2011 Chris Morgan <chmorgan@gmail.com>
+ */
+
+using System;
+using System.IO;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace SharpPcap.LibPcap
+{
+    /// <summary>
+    /// Per http://msdn.microsoft.com/en-us/ms182161.aspx 
+    /// </summary>
+    internal static partial class LibPcapSafeNativeMethods
+    {
+
+        [DllImport("kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        static extern bool SetDllDirectory(string lpPathName);
+
+        static LibPcapSafeNativeMethods()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                SetDllDirectory(Path.Combine(Environment.SystemDirectory, "Npcap"));
+            }
+            else
+            {
+                RegisterResolver();
+            }
+        }
+
+        /// <summary>
+        /// The class NativeLibrary is only available since .NET Core 3.0
+        /// It's not available in .NET Framework, 
+        /// but that does not affect us since we would use the Windows dll name wpcap
+        /// </summary>
+        private static void RegisterResolver()
+        {
+            var nativeLibraryType = typeof(DllImportSearchPath).Assembly
+                .GetType("System.Runtime.InteropServices.NativeLibrary");
+
+            if (nativeLibraryType == null)
+            {
+                return;
+            }
+
+            var dllImportResolverType = typeof(DllImportSearchPath).Assembly
+                .GetType("System.Runtime.InteropServices.DllImportResolver");
+
+            var setDllImportResolverMethod = nativeLibraryType
+                .GetMethod(
+                    "SetDllImportResolver",
+                    BindingFlags.Public | BindingFlags.Static,
+                    null,
+                    new[] { typeof(Assembly), dllImportResolverType },
+                    null
+                );
+
+
+            var dllImportResolver = Delegate.CreateDelegate(
+                dllImportResolverType, 
+                typeof(LibPcapSafeNativeMethods).GetMethod(nameof(Resolver))
+            );
+
+            setDllImportResolverMethod.Invoke(null, new object[] { 
+                typeof(LibPcapSafeNativeMethods).Assembly,
+                dllImportResolver
+            });
+        }
+
+        [DllImport("libdl")]
+        private static extern IntPtr dlopen(string filename, int flags);
+
+        public static IntPtr Resolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
+        {
+            if (libraryName != PCAP_DLL)
+            {
+                // Use default resolver
+                return IntPtr.Zero;
+            }
+
+            const int RTLD_NOW = 2;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                return dlopen("libpcap.so", RTLD_NOW);
```

Comment:
we need to look at adding some variants here. Some distros don't provide a libpcap.so file, but rather a libpcap.so.X or libpcap.so.X.Y even. It's been an issue in the past with the mono dllmap. Is this going to search for any library with a root of 'libpcap.so' or just that one library name?

Reply:
* I assume will only work for `libpcap.so`, nothing in [dlopen](https://linux.die.net/man/3/dlopen) about prefix.
* If you have a list of known `libpcap.so.X.Y` filenames, we could add the code to iterate over them and return the first found one.

Reply:
So in ubuntu they have a symlink but no raw libpcap.so library.

$ find -name "*pcap*so*"|xargs ls -l
lrwxrwxrwx 1 root root     16 Mar 23  2020 ./x86_64-linux-gnu/libpcap.so.0.8 -> libpcap.so.1.9.1
-rw-r--r-- 1 root root 302424 Mar 23  2020 ./x86_64-linux-gnu/libpcap.so.1.9.1

Reply:
@chmorgan can we merge this PR as is, and handle the versioning of so files in a different PR?

Reply:
@kayoub5 that works, its going to be a big win to handle those different filenames, its a huge pita today for some distros